src/messages.cpp

```diff
@@ -238,159 +241,235 @@ bool Messages::has_undisplayed_messages()
+     *
+     * In case of NEWEST-TOP, the 'i' variable easily enough corresponds to the newest message.
+     * In case of OLDEST-TOP, the 'i' variable must be flipped- meaning the highest value of 'i' returns the result for the lowest value of 'i', etc.
+     * To achieve this, the 'flip_message' variable is set to either the value of 'msg_count', or '0'. This variable is then subtracted from 'i' in each call to player_messages.impl_->history();
+     *
+     * 'offset' refers to the corresponding message that will be displayed at the very TOP of the message box window.
+     *  NEWEST-TOP: 'offset' starts simply at '0' - the very top of the window.
+     *  OLDEST-TOP: 'offset' is set to the maximum value it could possibly be. That is- 'msg_count-bottom'. This way, the screen starts with the scrollbar all the way down.
+     * 'retrieve_history' refers to the line that should be displayed- this is either 'i' if it's NEWEST-TOP, or a flipped version of 'i' if it's OLDEST-TOP.
+     */
+    int offset = log_from_top ? 0 : ( msg_count - bottom );
+    const int flip = log_from_top ? 0 : msg_count - 1;
+    int retrieve_history;
+
+    for( ;; ) {
+        werase( w );
+        draw_border( w );
+        mvwprintz( w, bottom + 1, 32, c_red, _("Press %s to return"), ctxt.get_desc("QUIT").c_str() );
+        draw_scrollbar( w, offset, bottom, msg_count, 1, 0, c_white, true );
 
         int line = 1;
         int lasttime = -1;
         for( int i = offset; i < msg_count; ++i ) {
-            if (line > bottom) {
+            retrieve_history = abs( i - flip );
+            if( line > bottom ) {
                 break;
+                // This statement makes it so that no non-existent messages are printed (which usually results in a segfault)
+            } else if( retrieve_history >= msg_count ) {
+                continue;
             }
 
-            const game_message &m     = player_messages.impl_->history(i);
-            const nc_color col        = msgtype_to_color( m.type, false );
+            const game_message &m     = player_messages.impl_->history( retrieve_history );
             const calendar timepassed = calendar::turn - m.timestamp_in_turns;
-
-            if (timepassed.get_turn() > lasttime) {
-                mvwprintz(w, line++, 3, c_ltblue, _("%s ago:"),
-                    timepassed.textify_period().c_str());
+            const char *long_ago      = timepassed.textify_period().c_str();
+            nc_color col              = msgtype_to_color( m.type, false );
+
+            // Here we seperate the unit and amount from one another so that they can be properly padded when they're drawn on the screen.
+            // Note that the very first character of 'unit' is often a space (except for languages where the time unit directly follows the number.)
+            char unit[0];
+            int amount;
+            sscanf( long_ago, "%d%s", &amount, unit );
```

Comment:
This should invoke UB again: `sscanf` is instructed to write a string ("%s") into the char array `unit`, which happens to have no space for it (not even the terminating 0-character can be written there).

I suggest you separate amount and units with something like
```C++
const size_t a = long_ago.find_first_not_of( "0123456789" );
amount = std:atoi( long_ago.substr( 0, a ).c_str() );
std::string unit = long_ago.substr( a );
```